aesd-char-driver/main.c

```diff
@@ -199,25 +208,28 @@ ssize_t aesd_write(struct file *filp, const char __user *buf, size_t count,
 	wr_entry.buffptr = buffptr;
 	wr_entry.size = (total_buf_size != 0) ? total_buf_size:count;//handle partial data if any
 	
-	//take lock before writting the data into circular buffer
-	if (mutex_lock_interruptible(&dev->buf_lock)) {
-	    kfree(buffptr);
-            return 0;//-ERESTARTSYS;
-	}
-
 	rm_buffptr = aesd_circular_buffer_add_entry(&dev->buffer, &wr_entry);
 	mutex_unlock(&dev->buf_lock);
 
         if (rm_buffptr != NULL) { //overwritten entry, free it
 	   kfree(rm_buffptr);
 	}
 
+	*f_pos = *f_pos + ((total_buf_size != 0) ? total_buf_size:count);//TODO:Come back and check this
 	retval = count;
     } else { //partial data
+
+    	//take lock before accessing the part_mem_ptr
+	if (mutex_lock_interruptible(&dev->buf_lock)) {
+    	   return 0;//-ERESTARTSYS;
```

Comment:
@maheshchnt I would return an error here like `-ERESTARTSYS`, not sure why that's commented out.